semgrep/semgrep/rule_match.py

```diff
@@ -192,19 +160,98 @@ def _get_previous_line(self) -> str:
+        code = code.strip()
+        return code
+
+    @cli_unique_key.default
+    def get_cli_unique_key(self) -> Tuple:
         return (
-            self.id == other.id
-            and self.path == other.path
-            and self.lines_hash == other.lines_hash
-            and self.lines == other.lines
+            self.rule_id,
+            self.path,
+            self.start.offset,
+            self.end.offset,
+            self.message,
         )
 
+    @ci_unique_key.default
+    def get_ci_unique_key(self) -> Tuple:
+        return (self.rule_id, self.path, self.syntactic_context, self.index)
+
+    @ordering_key.default
+    def get_ordering_key(self) -> Tuple:
+        return (self.path, self.start, self.end, self.rule_id, self.message)
+
+    @syntactic_id.default
+    def get_syntactic_id(self) -> str:
+        # Upon reviewing an old decision,
+        # there's no good reason for us to use MurmurHash3 here,
+        # but we need to keep consistent hashes so we cannot change this easily
+        hash_int = pymmh3.hash128(str(self.ci_unique_key))
+        hash_bytes = int.to_bytes(hash_int, byteorder="big", length=16, signed=False)
+        return str(binascii.hexlify(hash_bytes), "ascii")
+
     @property
     def uuid(self) -> UUID:
-        return UUID(
-            bytes=hashlib.sha256(
-                str((self.id, self.path, self.lines_hash)).encode()
-            ).digest()[:16]
-        )
+        return UUID(hex=self.syntactic_id)
+
+    def __hash__(self) -> int:
+        return hash(self.cli_unique_key)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, type(self)):
+            return False
+        return self.cli_unique_key == other.cli_unique_key
+
+    def __lt__(self, other: "RuleMatch") -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self.ordering_key < other.ordering_key
+
+
+class RuleMatchSet(Set[RuleMatch]):
+    """
+    A set type which is aware
+    that two rule matches are not to be considered the same
+    even if they have the same line of code.
+    It amends rule matches that have identical code during insertion
+    to set a unique zero-indexed "index" value on them.
+    """
+
+    def __init__(self, __iterable: Optional[Iterable[RuleMatch]] = None) -> None:
+
+        self._seen_ci_keys: Set[str] = set()
+        if __iterable is None:
+            super().__init__()
+        else:
+            super().__init__(__iterable)
+
+    def add(self, rule_match: RuleMatch) -> None:
+        """
+        Add finding, even if the same (rule, path, code) existed.
+        This is used over regular `.add` to increment the finding's index
+        if it already exists in the set, thereby retaining multiple copies
+        of the same (rule_id, path, line_of_code) tuple.
+        """
+        while rule_match.ci_unique_key in self._seen_ci_keys:
+            rule_match = evolve(rule_match, index=rule_match.index + 1)
```

Comment:
Should we not just keep a mapping from ci_unique_key without the index -> number of things with that key so this operation is constant time instead of O(num collisions)?

Reply:
makes sense, this code I copied from semgrep-agent and then added more stuff to it, not realizing there's an easier way like you describe

Reply:
I'll just use a collections.Counter